pkg/billing/model.go

```diff
@@ -4,6 +4,34 @@ import (
 	"time"
 )
 
+type Event[Self any] interface {
+	AbsoluteEvent | IncrementalEvent
+
+	eventMethods[Self]
+}
+
+// eventMethods exists because Go does not let you assume that a value of generic type "Foo | Bar"
+// has the fields common to both Foo and Bar. So we still have to thread through these fields.
+//
+// The methods in this interface *could* have directly been a part of Event, but then we couldn't do
+// the "implements eventMethods" assertions below (because you aren't allowed to store a value of
+// type "Foo | Bar"; it can only be used as a constraint for generics).
+//
+// eventMethods also takes Self as a parameter because we want Event[...] to be implemented for the
+// raw AbsoluteEvent and IncrementalEvent types, not pointers to them, because that would make
+// things just a bit more annoying. But we still need to set field values, which we can ONLY do by
```

Comment:
I think maybe everything should just be a pointer maybe?

Reply:
Yeah... fixed it so we're using pointers now. Kind of sad, but it's a lot simpler, and ultimately for the best I think.